Add S3 Integration for LFS storage to p4-fusion

GSLakiZoli opened the pull request:
This pull request introduces S3 storage integration to the p4-fusion project via the AWS SDK for C++. The following major changes are included:

- Added AWS SDK as an external project using CMake’s ExternalProject_Add, building only the S3 component.
- Integrated S3 client functionality into the codebase, enabling upload operations toward S3 compatible servers.
- Refactored relevant modules to support S3 compatible uploads besides the already implemented LFS upload.
- Added the needed command line parameters to support S3 uploads.

These changes lay the foundation for S3 storage support for large files (LFS) in p4-fusion and improve the project’s extensibility for future cloud integrations.

Commit:
LFSClient constructor added for S3 server connection and command line parameters either

Commit:
Separate communication code into class hierarchy to handle the two supported communication modes (LFS, S3)

Commit:
Add aws-sdk dependency to the project

Commit:
Implement S3 upload

Commit:
Fix communication with MinIO server.

Commit:
Fix AWS SDK dependencies as find_library threw error on start of a brand new build

peter-esik commented:
I can see that now there are quite a few LFS-related source files, and you've put them into their own subfolder. Could you please give the same treatment to `lfs_client.<h|cc>`? E.g. `lfs` and `lfs/communication`

peter-esik commented on p4-fusion/CMakeLists.txt:
```diff
@@ -32,18 +34,61 @@ if (APPLE)
 
+include(ExternalProject)
+
+ExternalProject_Add(aws-sdk
```
I like this approach a lot. I've noticed that even a clean build won't delete and re'clone this. For the general case, that's good, because it takes quite a few minutes to build it. But if someone needs a "really" clean build, is there a way to achieve that besides manually deleting the aws-sdk folder?

GSLakiZoli replied:
I've introduced a build target named "deepclean" in cce8c24aa28e5c82c58ab3278ea2b45c6ab5a7d8 to delete the build and aws-sdk folders completely when needed.

peter-esik commented on p4-fusion/lfs_client.h:
```diff
@@ -16,18 +16,35 @@
 		Error
 	};
 
+	/**
+	 * Constructor for the LFSClient using LFS server API.
+	 */
 	LFSClient(GitAPI& gitAPI, const std::string& serverUrl, const std::string& username, const std::string& password, const std::vector<std::string>& lfsPatterns);
 
+	/**
+	 * Constructor for the LFSClient using S3 API.
+	 */
```
I don't think these comments add much value.

peter-esik replied:
Using dependency injection (requiring a Communicator instance as a ctor argument) here would be nice. In addition to the usual benefits:
* it would be clearer how to "invoke" native LFS vs. S3 behavior
* the username, pw, bucket etc. would be encapsulated in the corresponding Communicator derived class, clearing up a lot of "clutter" in `LFSClient`

GSLakiZoli replied:
I've fixed this in f9e4399f4f427d87b3f8633e1856226e9ffb32ee.

peter-esik commented on p4-fusion/lfscommunication/communicator.cc:
```diff
@@ -0,0 +1,14 @@
+std::unique_ptr<Communicator> Communicator::CreateS3(const std::string& serverURL, const std::string& bucket, const std::string& repository, const std::string& username, const std::string& password)
+{
+	return std::unique_ptr<S3Comm>(new S3Comm(serverURL, bucket, repository, username, password));
+}
```
Any practical advantages over creating these classes directly in client code?

GSLakiZoli replied:
I've removed them in f9e4399f4f427d87b3f8633e1856226e9ffb32ee.

peter-esik commented on p4-fusion/lfscommunication/communicator.cc:
```diff
@@ -0,0 +1,14 @@
+#include "communicator.h"
+#include "lfscomm.h"
+#include "s3comm.h"
+#include "openssl/sha.h"
```
This is unused

GSLakiZoli replied:
I've removed this file in f9e4399f4f427d87b3f8633e1856226e9ffb32ee.

peter-esik commented on p4-fusion/main.cc:
```diff
@@ -140,18 +143,41 @@ int Main(int argc, char** argv)
+			ERR("Unsupported lfsAPI type '" << lfsAPI << "'. Supported types are 'lfs' and 's3'.");
+			return 1;
+		}
+
```
We already have LFS-related arg validation in `arguments.cc` (`Arguments::IsValid`). I think this should be moved there.

peter-esik replied:
(By "this" I meant the validation parts.)

GSLakiZoli replied:
I've fixed the validation in df003cedd357b1a8fa80d79b80b21a733bbd4ba0.

peter-esik commented on p4-fusion/lfs_client.cc:
```diff
@@ -403,16 +40,7 @@ std::vector<char> LFSClient::CreatePointerFileContents(const std::vector<char>&
-		snprintf(hex, sizeof(hex), "%02x", hash[i]);
-		hexHash += hex;
-	}
+	std::string hexHash = CalcOID(fileContents);
```
Could be `const`

GSLakiZoli replied:
I've fixed this in df003cedd357b1a8fa80d79b80b21a733bbd4ba0.

peter-esik commented on p4-fusion/lfs_client.cc:
```diff
@@ -436,32 +64,30 @@ LFSClient::UploadResult LFSClient::UploadFile(const std::vector<char>& fileConte
 	}
 
-	if (!batchResponse.verifyUrl.empty())
+	if (!comm)
```
I don't see how this can ever realistically happen. Getting rid of this check would result in cleaner code, and there would be no need for the `UnknownAPI` error code.

GSLakiZoli replied:
I've removed the UnknownAPI error code in f9e4399f4f427d87b3f8633e1856226e9ffb32ee.

peter-esik commented on p4-fusion/lfs_client.cc:
```diff
@@ -436,32 +64,30 @@ LFSClient::UploadResult LFSClient::UploadFile(const std::vector<char>& fileConte
 	{
-		return uploadResult;
+	case API::LFS:
+		comm = Communicator::CreateLFS(m_ServerUrl, m_Username, m_Password);
+		break;
+	case API::S3:
+		comm = Communicator::CreateS3(m_ServerUrl, m_Bucket, m_Repository, m_Username, m_Password);
```
These classes don't seem to contain any per-upload state. Why is it necessary to create a new instance every time a file needs to be uploaded?

I think we should create a single instance in the constructor, and store that instead of the `API` enum. That will be more performant and also more abstract.

GSLakiZoli replied:
I've fixed the count of instances in f9e4399f4f427d87b3f8633e1856226e9ffb32ee.

peter-esik commented on p4-fusion/lfs_client.cc:
```diff
@@ -5,388 +5,12 @@
 #include "lfs_client.h"
-
 #include "openssl/sha.h"
-#include <curl/curl.h>
```
Regardless of other changes, I really like that the bulk of the native-LFS logic was moved to its own file!

lakizoli committed:
Fix LFS directory structure of the project

GSLakiZoli commented:
I've fixed the directory structure of the project in the 1571fc515006ae0256383022b88590c290af6031 change.

lakizoli committed:
Fix creation of communicator instances to remove some glue code.
Fix count of communicator instances as they need to be created only once per run, not per upload.

lakizoli committed:
Fix validation of app args and communicator init.

lakizoli committed:
Introduce deepclean build target to clean AWS SDK upon desire.

peter-esik commented on p4-fusion/lfs/communication/s3comm.cc:
```diff
@@ -0,0 +1,57 @@
+	object_request.SetBucket(m_Bucket);
+	object_request.SetKey(m_Repository + "/" + oid);
+
+	auto stream = Aws::MakeShared<Aws::StringStream>("", std::ios_base::in | std::ios_base::out | std::ios_base::binary);
+    stream->write(fileContents.data(), fileContents.size());
```
Hmm, sounds like a valid point.

According to Copilot, there is machinery in the AWS SDK to achieve this:

```
auto streamBuf = Aws::MakeShared<Aws::Utils::Stream::PreallocatedStreamBuf>("", 
		reinterpret_cast<unsigned char*>(const_cast<char*>(fileContents.data())), 
		fileContents.size());
	auto stream = Aws::MakeShared<Aws::IOStream>("", streamBuf.get());
	stream->seekg(0);
```

GSLakiZoli replied:
I've fixed this in 45ad27f64b99262b0b059bd35f8689d679e54695.

peter-esik commented on p4-fusion/lfs/communication/s3comm.cc:
```diff
@@ -0,0 +1,57 @@
+
+	bool isHttps = m_ServerURL.find("https://") == 0;
+
+    Aws::S3::S3ClientConfiguration config;
+    config.endpointOverride = m_ServerURL;
+    config.scheme = isHttps ? Aws::Http::Scheme::HTTPS : Aws::Http::Scheme::HTTP;
+    config.verifySSL = isHttps;
+    config.region = "us-east-1"; // MinIO ignores region, but AWS SDK requires it
+	config.useVirtualAddressing = false;
```
A quick glance at the documentation suggests that there is `AWS_REGION` and also a global config file. If one of those are present, maybe this `S3ClientConfiguration` picks those up already? If so, we should check if it's set, and leave it alone if it is.

I know our use case does not involve using AWS S3 directly, but this seems like a low-hanging fruit.

GSLakiZoli replied:
I've fixed this in 45ad27f64b99262b0b059bd35f8689d679e54695.

peter-esik commented on p4-fusion/lfs/communication/s3comm.cc:
```diff
@@ -0,0 +1,57 @@
+		OptionsInitializer(Aws::SDKOptions& opts) : options(opts) { Aws::InitAPI(opts);}
+		~OptionsInitializer() { Aws::ShutdownAPI(options); }
+	} optionsInitializer (options);
+
```
Sound like a fair point. Furthermore, the `S3Client` and `S3ClientConfiguration` instances I think can be reused, too. If not for performance, this would clean `S3Comm::UploadFile ()` up a bit.

GSLakiZoli replied:
I've fixed this in 45ad27f64b99262b0b059bd35f8689d679e54695.

lakizoli committed:
Fix AWS SDK initialization.

lakizoli committed:
Make the initialization thread safe.

GSLakiZoli commented on p4-fusion/lfs/communication/s3comm.cc:
```diff
@@ -0,0 +1,71 @@
+	Aws::SDKOptions awsOptions;
+};
+
+std::unique_ptr<AWSInitializer> awsInitializer;
```
I've fixed this in 0458241c4c50124204bc4d3be15bccdb649250e3.

peter-esik replied:
There is no thread safety concern here, as there is only one `S3Comm` instance created by the main thread (in `main`). But I guess it's no harm in making it thread safe.

GSLakiZoli commented on p4-fusion/lfs/communication/s3comm.cc:
```diff
@@ -0,0 +1,71 @@
+	object_request.SetBucket(m_Bucket);
+	object_request.SetKey(m_Repository + "/" + oid);
+
+	auto streamBuf = Aws::MakeShared<Aws::Utils::Stream::PreallocatedStreamBuf>("",
+		reinterpret_cast<unsigned char*>(const_cast<char*>(fileContents.data())), 
+		fileContents.size());
```
The copy was the original solution, so I would ignore this.

Commit:
Move AWS SDK init/release to the main function, because it is sensitive to the stack level of init/shutdown pair. The shutdown has to be run precisely at the same stack level, where start was run.

peter-esik commented on p4-fusion/lfs/communication/s3comm.cc:
```diff
@@ -0,0 +1,54 @@
+	Aws::S3::S3Client s3_client(credentials, config,
+        Aws::Client::AWSAuthV4Signer::PayloadSigningPolicy::Never, false);
+
+	Aws::S3::Model::PutObjectRequest object_request;
```
A file with the very same content can appear in multiple changelists. If this happens, we will upload the same file more than once. This provides an optimization opportunity: before uploading a file, we can query whether it was already uploaded or not. Interestingly, LFS' basic transfer/batch API supports this out-of-the box, and `p4-fusion` already makes use of this.

I've made some measurements with converting the BIMx codebase to see whether this would be worth doing, or not. The results are quite convincing:
* conversion time: 27m vs. 31m (13% improvement)
* upload bandwidth saved: 1.97 GiB (only logged the savings, so I don't know the baseline, sorry!)

The code to achieve this is quite simple:

```suggestion
	// Check if object already exists
	Aws::S3::Model::HeadObjectRequest headRequest;
	head_request.SetBucket(m_Bucket);
	head_request.SetKey(objectKey);

	auto headResult = s3_client.HeadObject(headRequest);
	if (headResult.IsSuccess()) {
		return UploadResult::AlreadyExists;
	}
```

peter-esik replied:
My suggestion is to add this piece of code, and not to remove this line, messed up the "Add code suggestion" thingy.

GSLakiZoli replied:
I've fixed this in b4fe3c24ad518656b80c693a6279ac1d0232beff.

peter-esik commented on p4-fusion/main.cc:
```diff
@@ -617,6 +637,9 @@ int main(int argc, char** argv)
 {
 	int exitCode = 0;
 
+	Aws::SDKOptions awsOptions;
+	Aws::InitAPI(awsOptions);
```
Just a nitpick, but all the logic is inside `Main`, and not `main`. It'd be nice to keep it that way.

GSLakiZoli replied:
I deliberately placed this change into main, because this is the only single point, where the shutdown in success case and in every error case can be handled super easy. In the Main there should be some Automatic remover struct introduced to achieve this and morover it not handles the exception handling case implemented in main. So i think placing this in main instead of Main helps the readability well.

peter-esik commented on p4-fusion/lfs/communication/s3comm.cc:
```diff
@@ -0,0 +1,54 @@
+    config.region = Aws::Region::US_EAST_1; // MinIO ignores region, but AWS SDK requires it
+	config.useVirtualAddressing = false;
+
+	Aws::S3::S3Client s3_client(credentials, config,
```
Could these be initialized once in the ctor, instead of each time we upload a file?

GSLakiZoli replied:
Yeah, it could, but in that case the S3Comm class would contain AWS dependencies on its interface. Hence, all of its dependencies would depend on AWS either. I think that would be slightly worse than recreating the credentials and config before every upload. In practice this runs very fast, so separating this code is not worth the effort.

peter-esik replied:
Sounds fair.

peter-esik commented on p4-fusion/lfs/communication/s3comm.cc:
```diff
@@ -0,0 +1,54 @@
+    config.endpointOverride = m_ServerURL;
+    config.scheme = isHttps ? Aws::Http::Scheme::HTTPS : Aws::Http::Scheme::HTTP;
+    config.verifySSL = isHttps;
+    config.region = Aws::Region::US_EAST_1; // MinIO ignores region, but AWS SDK requires it
```
Are you sure there is no default value here? The comments say: `Default US_EAST_1`

Furthermore, env. vars and a global config file (`~/.aws/config`) can also declare options, and this class picks them up. Not setting (overwriting) anything unless necessary would be nice.

GSLakiZoli replied:
I've removed this in b4fe3c24ad518656b80c693a6279ac1d0232beff.

Commit:
Introduce HEAD query to not upload files already stored in the repo.
Remove region definition as it is the default value in the AWS and MinIO ignores the value completely.
Rename local variables according to style guide.

Commit:
Merge changes from gs-specific branch to gs-s3 branch

mmarczell-graphisoft committed:
Build fix

lakizoli committed:
S3 bucket name validation implemented.
